compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -254,14 +271,18 @@ impl ResolverAstLoweringExt for ResolverAstLowering {
 
 /// Context of `impl Trait` in code, which determines whether it is allowed in an HIR subtree,
 /// and if so, what meaning it has.
-#[derive(Debug, Copy, Clone, PartialEq, Eq)]
-enum ImplTraitContext {
+#[derive(Debug)]
+enum ImplTraitContext<'b, 'itctx> {
     /// Treat `impl Trait` as shorthand for a new universal generic parameter.
     /// Example: `fn foo(x: impl Debug)`, where `impl Debug` is conceptually
     /// equivalent to a fresh universal parameter like `fn foo<T: Debug>(x: T)`.
     ///
     /// Newly generated parameters should be inserted into the given `Vec`.
-    Universal,
+    Universal {
+        apit_nodes: &'b mut Vec<&'itctx Ty>,
```

Comment:
Can't this be put in `LoweringContext`, with `impl_trait_defs` and `impl_trait_bounds` since they are related?

Reply:
We (me and @nikomatsakis) were trying to do the opposite, actually. Trying to have more local data instead of growing `LoweringContext` more and more, which already is like a god object.
Anyway, as I've commented elsewhere, I think we should remove `apit_nodes` in favor of having a visitor that gathers this information. If you're ok, I'd leave for a followup.